### DISCUSSION TITLE
Why ckb uses tentacle (nervosnetwork/p2p) instead of libp2p?

### DISCUSSION THREAD

**Discussion by doitian** [nervosnetwork/ckb#2429] [General]
See the title.

**Comment by zhaxzhax** [comment]
Same question.  The newest p2p version is publish in 2018. Looks like crate p2p is abandoned.  Meanwhile,  lib-p2p is still being updated.

**Comment by doitian** [comment]
> Same question. The newest p2p version is publish in 2018. Looks like crate p2p is abandoned. Meanwhile, lib-p2p is still being updated.

CKB uses [nervosnetwork/p2p](https://github.com/nervosnetwork/p2p) not the crate [p2p](https://crates.io/crates/p2p).

**Comment by zhaxzhax** [comment]
I have heard that cause of  [driftluo meet some magic bugs](https://www.driftluo.com/article/da927b75-47b6-4fcd-8289-4a373e46331c) when used libp2p,  he choose to build a another crate for easily using.

I have the experience using _libp2p_, its plenty of trait made it really hard to build a simple application ( complex api ). Most important, _libp2p_ not support wasm, since it depend on crate _net2_.
So here is my question, does tentacle support wasm or will tentacle support wasm one day?

**Comment by doitian** [comment]
There's no plan to support wasm yet. But @driftluo [is implementing the protocol in Go](https://github.com/driftluo/tentacle-go). If it supports wasm target, then we'll have a wasm implementation as well.

**Comment by driftluo** [comment]
@zhaxzhax tentacle support wasm now
